doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,443 @@
+
+Before we can propose alternatives to the current datatype system,
+it is helpful to review how "value based promotion" is used and can be useful.
+Value based promotion allows for the following code to work::
+
+    # Create uint8 array, as this is sufficient:
+    uint8_arr = np.array([1, 2, 3], dtype=np.uint8)
+    result = uint8_arr + 4
+    result.dtype == np.uint8
+
+    result = uint8_arr * (-1)
+    result.dtype == np.int16  # upcast as little as possible.
+
+Where especially the first part can be useful: The user knows that the input
+is an integer array with a specific precision. Considering that plain ``+ 4``
+retaining the previous datatype is intuitive.
+Replacing this example with ``np.float32`` is maybe even more clear,
+as float will rarely have overflows.
+Without this behaviour, the above example would require writing ``np.uint8(4)``
+and lack of the behaviour would make the following suprising::
+
+    result = np.array([1, 2, 3], dtype=np.float32) * 2.
+    result.dtype == np.float32
+
+where lack of a special case would cause ``float64`` to be returned.
+
+It is important to note that the behaviour also applies to universal functions
+and zero dimensional arrays::
+
+    # This logic is also used for ufuncs:
+    np.add(uint8_arr, 4).dtype == np.uint8
+    # And even if the other array is explicitly typed:
+    np.add(uint8_arr, np.array(4, dtype=np.int64)).dtype == np.uint8 
+
+To review, if we replace ``4`` with ``[4]`` to make it one dimensional, the
+result will be different::
+
+    # This logic is also used for ufuncs:
+    np.add(uint8_arr, [4]).dtype == np.int64  # platform dependend
+    # And even if the other array is explicitly typed:
+    np.add(uint8_arr, np.array([4], dtype=np.int64)).dtype == np.int64
+
+
+Proposed Weak Promotion
+-----------------------
+
+This proposal uses a "weak scalar" logic.  This means that Python ``int``, ``float``,
+and ``complex`` are not assigned one of the typical dtypes, such as float64 or int64.
+Rather, they are assigned a special abstract DType, similar to the "scalar" hierarchy
+names: Integral, Floating, ComplexFloatin.
+
+When promotion occurs (as it does for ufuncs if no exact loop matches),
+the other DType is able to decide how to regard the Python
+scalar.  E.g. a ``UInt16`` promoting with an `Integral` will give ``UInt16``.
+
+.. note::
+
+    A default will most likely be provided in the future for user defined DTypes.
+    Most likely this will end up being the default integer/float, but in principle
+    more complex schemes could be implemented.
+
+At no time is the value used to decide the result of this promotion.  The value is only
+considered when it is converted to the new dtype; this may raise an error.
+
+
+
+
+Related Work
+============
+
+* `JAX promotion`_ also uses the weak-scalar concept.  However, it makes use
+  of it also for most functions.  JAX further stores the "weak-type" information
+  on the array: ``jnp.array(1)`` remains weakly typed.
```

Comment:
Heh, the JAX promotion docs are pretty detailed. I was thinking we might want to briefly mention if our new/proposed direction was consistent with some other major languages/frameworks, though not sure if that changes much.

Reply:
Yeah, JAX uses the weak logic, I think I should be reporting some bugs to them, but I don't remember (I believe if you swapped the inputs on some of their example, things got wrong, and JAX tags the "was scalar" to the 0-D array, so it can be passed around, which I am not sure always turns out right, but is easier for them because their arrays are immutable.

I believe most use "weak" logic (torch definitely does), although all of these skip scalars entirely probably.